### PR TITLE
Big clean up, more details in the PR

### PR DESCRIPTION
**Pull request by shedaniel**
- Add ArchitecturySpawnEggItem, Similar with ForgeSpawnEggItem, but does not duplicate the vanilla code as much. Uses the new `RegistrySupplier#listen`
- Unify AWs, switch more hooks to transitive AWs:
  - GameRuleFactory
  - GameRuleRegistry
  - CriteriaTriggersRegistry
- Remove All `@NotNull`
- Add `getRegistries` and `getRegistrar` in `DeferredRegister` and `RegistrySupplier`
- Add `wrap` in `Registrar` to wrap already existing entries as `RegistrySupplier`s
- Add `listen` in `Registrar`
  - Fabric: Uses `RegistryEntryAddedCallback`
  - Forge Registry: Callback at our `register` or at post-`RegistryEvent.Register`
  - Forge Vanilla Registry: Callback at `LifecycleEvent.SETUP`, there isn't a hook for listening to added registry entries, we might mixin ourselves
- Deprecate `BlockProperties`

### CODE REVIEW

File: fabric/src/main/java/dev/architectury/registry/registries/fabric/RegistriesImpl.java [hunk 1]
```diff
@@ -79,27 +99,50 @@
...
         
         @Override
-        @NotNull
         public <T> RegistrarBuilder<T> builder(Class<T> type, ResourceLocation registryId) {
-            return new RegistrarBuilderWrapper<>(FabricRegistryBuilder.createSimple(type, registryId));
+            return new RegistrarBuilderWrapper<>(modId, FabricRegistryBuilder.createSimple(type, registryId));
+        }
+    }
+    
+    public static class RegistryEntryId<T> {
```
Comment (MaxNeedsSnacks) [inline fabric/src/main/java/dev/architectury/registry/registries/fabric/RegistriesImpl.java:107]: This could just be a record, couldn't it?

Comment (shedaniel) [reply fabric/src/main/java/dev/architectury/registry/registries/fabric/RegistriesImpl.java:107]: Wanted easier port back to 1.16

Comment (MaxNeedsSnacks) [reply fabric/src/main/java/dev/architectury/registry/registries/fabric/RegistriesImpl.java:107]: Fair enough, might be worth keeping in mind as a post-impl note for 1.18 tho

**Review by MaxNeedsSnacks — APPROVED** [review]

File: common/build.gradle [hunk 2]
```diff
@@ -12,6 +12,108 @@ architectury {
...
+            .map { loadClass(it) }
+            .sorted(Comparator.comparing { it.name })
+            .filter { (it.access & Opcodes.ACC_ABSTRACT) == 0 }
+            .forEach { node ->
+                for (def method : node.methods) {
+                    // Checklist for finding block constructors as of 1.18.2:
+                    //  - class directly in net.minecraft.world.level.block (excluding subpackages)
+                    //  - method name == <init> (by definition)
+                    //  - contains an BlockBehaviour$Properties parameter
+                    //  - only taking into account non-abstract classes and non-public constructors
```
Comment (Juuxel) [inline common/build.gradle:74]: please credit me for this code, it's literally copy-pasted and changed to mojmap

Comment (shedaniel) [reply common/build.gradle:74]: oh right I forgot

**Review by MaxNeedsSnacks — APPROVED** [review]

**Review by Juuxel — APPROVED** [review]

**Pull request merged**